pkg/roachprod/failureinjection/failures/disk_stall.go

```diff
@@ -0,0 +1,474 @@
+) (int, int, error) {
+	maj, min, err := s.DiskDeviceMajorMinor(ctx, l)
+	if err != nil {
+		return 0, 0, err
+	}
+	// Check the number of bytes read and written to disk.
+	res, err := s.RunWithDetails(
+		ctx, l, node,
+		fmt.Sprintf(`grep -E '%d:%d' /sys/fs/cgroup/system.slice/io.stat |`, maj, min),
+		`grep -oE 'rbytes=[0-9]+|wbytes=[0-9]+' |`,
+		`awk -F= '{printf "%s ", $2} END {print ""}'`,
+	)
+	if err != nil {
+		return 0, 0, err
+	}
+	fields := strings.Fields(res.Stdout)
+	if len(fields) != 2 {
+		return 0, 0, errors.Errorf("expected 2 fields, got %d: %s", len(fields), res.Stdout)
+	}
+
+	readBytes, err := strconv.Atoi(fields[0])
+	if err != nil {
+		return 0, 0, err
+	}
+	writeBytes, err := strconv.Atoi(fields[1])
+	if err != nil {
+		return 0, 0, err
+	}
+
+	return readBytes, writeBytes, nil
+}
+
+type DmsetupDiskStaller struct {
+	GenericFailure
+}
+
+func MakeDmsetupDiskStaller(
+	clusterName string, l *logger.Logger, secure bool,
+) (FailureMode, error) {
+	c, err := roachprod.GetClusterFromCache(l, clusterName, install.SecureOption(secure))
+	if err != nil {
+		return nil, err
+	}
+
+	genericFailure := GenericFailure{c: c, runTitle: DmsetupDiskStallName}
+	return &DmsetupDiskStaller{GenericFailure: genericFailure}, nil
+}
+
+func (s *DmsetupDiskStaller) Description() string {
+	return "dmsetup disk staller"
+}
+
+func (s *DmsetupDiskStaller) Setup(ctx context.Context, l *logger.Logger, args FailureArgs) error {
+	diskStallArgs := args.(DiskStallArgs)
+	var err error
+
+	// Disabling journaling requires the cockroach process to not have been started yet.
+	if diskStallArgs.RestartNodes {
+		// Use the default stop opts, if the user wants more control, they should manage
+		// the cluster restart themselves.
+		stopOpts := roachprod.DefaultStopOpts()
+		if err = s.StopCluster(ctx, l, stopOpts); err != nil {
+			return err
+		}
+	}
+
+	dev, err := s.DiskDeviceName(ctx, l)
+	if err != nil {
+		return err
+	}
+
+	// snapd will run "snapd auto-import /dev/dm-0" via udev triggers when
+	// /dev/dm-0 is created. This possibly interferes with the dmsetup create
+	// reload, so uninstall snapd.
+	if err = s.Run(ctx, l, s.c.Nodes, `sudo apt-get purge -y snapd`); err != nil {
+		return err
+	}
+	if err = s.Run(ctx, l, s.c.Nodes, `sudo umount -f /mnt/data1 || true`); err != nil {
+		return err
+	}
+	if err = s.Run(ctx, l, s.c.Nodes, `sudo dmsetup remove_all`); err != nil {
+		return err
+	}
+	// See https://github.com/cockroachdb/cockroach/issues/129619#issuecomment-2316147244.
+	if err = s.Run(ctx, l, s.c.Nodes, `sudo tune2fs -O ^has_journal `+dev); err != nil {
+		return errors.WithHintf(err, "disabling journaling fails if the cluster has been started")
+	}
+	if err = s.Run(ctx, l, s.c.Nodes, `echo "0 $(sudo blockdev --getsz `+dev+`) linear `+dev+` 0" | `+
+		`sudo dmsetup create data1`); err != nil {
+		return err
+	}
+	// This has occasionally been seen to fail with "Device or resource busy",
```

Comment:
Nit: missing TODO?

Reply:
This was copy and pasted from the existing disk stall code, i'm not too familiar with the context unfortunately.